src/core/gltf2importer/meshparser.cpp

```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+        return userData->indexAttribute->count() / 3;
+    };
+
+    interface.m_getNumVerticesOfFace = [](const SMikkTSpaceContext *pContext, const int iFace) ->int {
+        return 3;
+    };
+
+    interface.m_getPosition = [] (const SMikkTSpaceContext *pContext,
+                    float fvPosOut[],
+                    const int iFace,
+                    const int iVertex) {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
+        const auto vertexIndices = findVertexIndicesInFace(userData->indexAttribute, iFace);
+        const auto vertexIndex = vertexIndices[iVertex];
+        const auto vertexBuffer = userData->positionAttribute->buffer();
+        const auto byteOffset = userData->positionAttribute->byteOffset();
+        const auto byteStride = userData->positionAttribute->byteStride() == 0 ?
+                                3*4 : // Position vertex base type must be float
+                                userData->positionAttribute->byteStride();
+        const auto &vertexBufferData = vertexBuffer->data();
+        const auto positionHead = vertexBufferData.mid(byteOffset + vertexIndex*byteStride);
+        const auto *typedPositionHead = reinterpret_cast<const float*>(positionHead.data());
+        fvPosOut[0] = typedPositionHead[0];
+        fvPosOut[1] = typedPositionHead[1];
+        fvPosOut[2] = typedPositionHead[2];
+    };
+
+    interface.m_getNormal = [] (const SMikkTSpaceContext *pContext,
+                    float fvPosOut[],
+                    const int iFace,
+                    const int iVertex) {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
+        const auto vertexIndices = findVertexIndicesInFace(userData->indexAttribute, iFace);
+        const auto vertexIndex = vertexIndices[iVertex];
+        const auto vertexBuffer = userData->normalAttribute->buffer();
+        const auto byteOffset = userData->normalAttribute->byteOffset();
+        const auto byteStride = userData->normalAttribute->byteStride() == 0 ?
+                                3*4 : // Position vertex base type must be float
+                                userData->normalAttribute->byteStride();
+        const auto &vertexBufferData = vertexBuffer->data();
+        const auto positionHead = vertexBufferData.mid(byteOffset + vertexIndex*byteStride);
+        const auto *typedPositionHead = reinterpret_cast<const float*>(positionHead.data());
+        fvPosOut[0] = typedPositionHead[0];
+        fvPosOut[1] = typedPositionHead[1];
+        fvPosOut[2] = typedPositionHead[2];
+    };
+
+    interface.m_getTexCoord = [] (const SMikkTSpaceContext *pContext,
+                    float fvPosOut[],
+                    const int iFace,
+                    const int iVertex) {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
```

Comment:
Lots of duplication around those m_get*, also lots of the work is done more than once. I think we should aim at de-duplicating and reifying the computation in the factored code to make it faster. It's all based on already known addresses and properties of the various attributes at the creation of those lambda, let's lock the lambda behavior at that creation time.

In most of those the only missing information is vertexIndex, all the rest should be already known.

Reply:
So Im adding some helper structs that can be used to cache some searchs, have less duplication and make it faster